Refactor worker into staged pipeline

werkt opened the pull request:
Move the major operations - Input Fetch, Execute, Writeback, of the
worker into a pipeline with concurrent stage execution. The number of
available concurrent slots for the execution stage is configurable, and
each stage will stall on a transition to the next if it is (fully)
occupied.

ola-rozenfeld commented:
I just returned from my vacation, will do my best to review this today.

ola-rozenfeld commented on src/main/java/build/buildfarm/worker/ExecuteActionStage.java:
```diff
@@ -0,0 +1,117 @@
+
+  @Override
+  public PipelineStage error() {
+    return this.output;
```
this.error, presumably?

werkt replied:
removed code, good catch though!

ola-rozenfeld commented on src/main/java/build/buildfarm/worker/AbstractPipelineStage.java:
```diff
@@ -0,0 +1,116 @@
+  private final PipelineStage error;
+  private final BlockingQueue<OperationContext> queue;
+
+  private PipelineStage input;
```
It looks like you need all these variables in subclasses, so maybe add protected accessors?

werkt replied:
There's no usage of these, but that highlights that these are completely unnecessary except for Executor->ExecuteActionStage communication.  I've updated the diff to clean this up.

ola-rozenfeld commented:
Err, I did not realize this sends you comments right away -- sorry, my
first time using this review tool. My intent was to write a lot of drafts
and send them all at once.

On Mon, Jan 8, 2018 at 12:49 PM, Ola Rozenfeld <notifications@github.com>
wrote:

> *@ola-rozenfeld* commented on this pull request.
> ------------------------------
>
> In src/main/java/build/buildfarm/worker/AbstractPipelineStage.java
> <https://github.com/bazelbuild/bazel-buildfarm/pull/63#discussion_r160209691>
> :
>
> > +// distributed under the License is distributed on an "AS IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +package build.buildfarm.worker;
> +
> +import java.util.concurrent.BlockingQueue;
> +import java.util.concurrent.ArrayBlockingQueue;
> +
> +abstract class AbstractPipelineStage implements PipelineStage {
> +  private final PipelineStage output;
> +  private final PipelineStage error;
> +  private final BlockingQueue<OperationContext> queue;
> +
> +  private PipelineStage input;
>
> It looks like you need all these variables in subclasses, so maybe add
> protected accessors?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel-buildfarm/pull/63#pullrequestreview-87285578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYoKuOlHihEHdyL5fQ9ch2xyjcifbzGmks5tIlUhgaJpZM4QDOV2>
> .
>

werkt commented:
No problem, ulf had the same reaction - not sure if there's any other way with github review.  I can deal with it like this, or if you figure out how to have a bulk review come in, let me know how!

ola-rozenfeld commented on src/main/java/build/buildfarm/worker/ReportResultStage.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    ImmutableList.Builder<ByteString> contents = new ImmutableList.Builder<>();
+    if (resultBuilder.getExitCode() == 0) {
```
Wait, are you only uploading artifacts to CAS on success? Then, how will the user see logs of failed tests, for example?

werkt replied:
I was. This appears to be incorrect as we've discussed offline: regardless of exit status, all output files (that exist) should be read from and uploaded into the CAS at the completion of an action execution. Will adjust

ola-rozenfeld commented on src/main/java/build/buildfarm/worker/ReportResultStage.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    ActionResult result = resultBuilder.build();
+    if (!operationContext.action.getDoNotCache()) {
```
and if exit code is 0, right? You don't want to put failed actions into the cache.

werkt replied:
Likewise added this test. Mentioned that we may want to make this configurable, in cases where we want to trust the failure.  The bazel-supported configuration will only have this happen on success, correct.

ola-rozenfeld commented on src/main/java/build/buildfarm/worker/Executor.java:
```diff
@@ -0,0 +1,244 @@
+      }
+      if (control.getPolicy() == CASInsertionControl.Policy.ALWAYS_INSERT ||
+          (!withinLimit && control.getPolicy() == CASInsertionControl.Policy.INSERT_ABOVE_LIMIT)) {
+        resultBuilder.setStdoutDigest(worker.instance.putBlob(stdoutRaw));
```
So I guess this is an exception to the stage separation concept -- you upload stuff here from the "wrong" stage? I guess you could separate the stages completely, but then you'd need to save the stdout/err as "special" files, which might not be worth it, or pass them around in some other way.

werkt replied:
This is definitely an issue - I will have to transport these outputs through the context to the report result stage, good catch.

werkt commented on src/main/java/build/buildfarm/worker/Worker.java:
```diff
@@ -161,323 +101,27 @@ public void start() {
+    pipeline.add(matchStage, 1);
+    pipeline.add(inputFetchStage, 2);
+    pipeline.add(executeActionStage, 3);
+    pipeline.add(reportResultStage, 4);
```
Yes, exactly. From the pipeline standpoint, it doesn't make any sense for a stage to be running if there is nowhere for its output to go.

werkt commented on src/main/java/build/buildfarm/worker/Pipeline.java:
```diff
@@ -0,0 +1,96 @@
+        Thread thread = stageThread.getValue();
+        try {
+          thread.join(closeStage ? 1 : 1000);
+        } catch (InterruptedException ex) {
```
You have prompted me to investigate fully the ramifications of this.  According to the docs, and the observed behavior of Java, `InterruptedException` is thrown when the *current* thread is interrupted, which is the one calling .join (and not the object it is calling it on). It sounds an awful lot in your comment here that you're under the impression that `InterruptedException` is thrown on `Thread#join(...)` when the thread being joined with has been interrupted... (Can supply example with this exhaustedly proven false)

I'm on board with throwing an InterruptedException if we get same during this process, however I definitely don't want to bail out early, leaving main/the system exit to sit with unstopped attached threads (and hang forever).

werkt commented on src/main/java/build/buildfarm/worker/AbstractPipelineStage.java:
```diff
@@ -0,0 +1,108 @@
+  @Override
+  public synchronized void release() {
+    claimed = false;
+    notify();
```
This will not notify all threads, it will only notify one (the notifyAll method satisfies the all case). The notify is unspecified, which makes it a this-call, on the stage object, which makes that object itself a 'per-stage lock object' (that also happens to be the 'per-stage object')

werkt commented on src/main/java/build/buildfarm/worker/AbstractPipelineStage.java:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  @Override
+  public void close() {
```
Correct.

werkt commented on src/main/java/build/buildfarm/worker/ExecuteActionStage.java:
```diff
@@ -0,0 +1,115 @@
+
+  @Override
+  public synchronized void release() {
+    if (!executors.remove(Thread.currentThread())) {
```
Yes, that is exactly the pattern.

werkt commented on src/main/java/build/buildfarm/worker/Pipeline.java:
```diff
@@ -0,0 +1,96 @@
+
+        if (!thread.isAlive()) {
+          System.out.println("Stage has exited at priority " + stageClosePriorities.get(stage));
+          if (!stage.isClosed()) {
```
Yes, you're definitely right - this was either vestigial or paranoid. I would say that interrupted is not the only way for a stage thread to die, but in the normal sequence of operations, this should be sufficient.

werkt commented on src/main/java/build/buildfarm/worker/Pipeline.java:
```diff
@@ -0,0 +1,96 @@
+          }
+          inactiveStages.add(stage);
+        } else if (stage.isClosed()) {
+          System.out.println("Interrupting unterminated closed thread at priority " + stageClosePriorities.get(stage));
```
Yes, if they haven't reacted to being closed yet (i.e. are still in the middle of matching/fetching/executing/reporting)

werkt commented:
I think I've addressed all points except for the refactor of AbstractPipelineStage, just because I'm trying to normalize the behaviors of the Abstract, Match, and Execute stages (specifically their #iterate facsimiles) for a common interface. I won't leave it hanging, but need another day or so.

ola-rozenfeld commented:
Is this ready for another pass?

On Fri, Jan 12, 2018 at 3:19 PM, George Gensure <notifications@github.com>
wrote:

> I think I've addressed all points except for the refactor of
> AbstractPipelineStage, just because I'm trying to normalize the behaviors
> of the Abstract, Match, and Execute stages (specifically their #iterate
> facsimiles) for a common interface. I won't leave it hanging, but need
> another day or so.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel-buildfarm/pull/63#issuecomment-357332321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYoKuMKvXEMBrmNq38IH5qp6N77b4iesks5tJ73RgaJpZM4QDOV2>
> .
>

werkt commented:
> Is this ready for another pass?

Yes, please take a look.  AbstractPipelineStage is now PipelineStage.
